### PR TITLE
Added minimal DFA for finite language algorithm

### PR DESCRIPTION
**Pull request by eliotwrobson**
Added algorithm to compute minimal DFA from given finite language. Still a bit of a WIP, but should be good to go soon.

@caleb531 marked ready for review (since tests cases are passing and style is good), though I think the merge conflict is from changes you're making. Not sure if you wanted to make more changes to the base branch before I resolved the conflict.

### CODE REVIEW

File: tests/test_dfa.py [hunk 1]
```diff
@@ -1396,3 +1397,46 @@ def test_show_diagram_write_file(self):
         self.dfa.show_diagram(path=diagram_path)
         self.assertTrue(os.path.exists(diagram_path))
         os.remove(diagram_path)
+
+
+    def test_minimal_finite_language(self):
+        """Should compute the minimal DFA accepting the given finite language"""
+
+        # Same language described in the book this algorithm comes from
+        language = ['aa','aaa','aaba','aabbb','abaa','ababb','abbab','baa','babb','bbaa','bbabb','bbbab']
```
Comment (caleb531) [inline tests/test_dfa.py:1406]: @eliotwrobson Shouldn't this be a `set`, not a `list`?

Comment (eliotwrobson) [reply tests/test_dfa.py]: Good catch!

**Review by caleb531 — CHANGES_REQUESTED** [review]
Left one more request to convert the pytest `assert` statements to `self.assert*` methods.

As far as branching goes, I rebased your commits on top of my latest `immutability` branch. I also deleted `thing.py` before it got committed during the rebase. So you should be all set now as far as branch alignment.

File: tests/test_dfa.py [hunk 2]
```diff
@@ -1396,3 +1397,46 @@ def test_show_diagram_write_file(self):
...
+                8: {'a': 0, 'b': 9},
+                9: {'a': 0, 'b': 0},
+            },
+            initial_state=1,
+            final_states={4, 9}
+        )
+
+        minimal_dfa = DFA.from_finite_language(language, {'a', 'b'})
+
+        assert len(minimal_dfa.states) == len(equiv_dfa.states)
```
[the remark this answers is not in the document]
Comment (caleb531) [reply tests/test_dfa.py]: @eliotwrobson Actually, while I have your fork open, let me take care of this one. 🙂

**Review by caleb531 — APPROVED** [review]
Took the liberty of fixing the pytest assertions, and I already resolved the branching, so this should be all good now.

**Pull request merged**